### PR TITLE
RSA keys set to PSS/OAEP padding perform PKCS1v1.5 when MBEDTLS_USE_PSA_CRYPTO is enabled

### LINKED ISSUE #8824: RSA keys set to PSS/OAEP padding perform PKCS1v1.5 when MBEDTLS_USE_PSA_CRYPTO is enabled
In the default configuration, if you have an `mbedtls_pk_context` of type `MBEDTLS_PK_RSA`, and the underlying RSA context has its padding mode set to `MBEDTLS_RSA_PKCS_V21`, then the functions `mbedtls_pk_sign`, `mbedtls_pk_verify`, `mbedtls_pk_encrypt` and `mbedtls_pk_decrypt` perform PSS or OAEP. If the padding mode is left at the default `MBEDTLS_RSA_PKCS_V21`, then those functions perform a PKCS#1v1.5 algorithm. This has been the case since the PolarSSL days.

(There are functions `mbedtls_pk_sign_ext` and `mbedtls_pk_verify_ext` that bypass the underlying context's padding mode, but they aren't relevant to the behavior of the non-ext function.)

This is not clearly documented, but it's a reasonable expectation. The documentation of the pk functions sttate that “For RSA keys, the default padding type is PKCS#1 v1.5.” That's the default if you construct a key with `mbedtls_pk_parse_xxx`, but if you manipulate the underlying RSA context, it changes the algorithm.

Bug: if you enable `MBEDTLS_USE_PSA_CRYPTO`, then since #5519 released in Mbed TLS 3.2, the PK module dispatches to PSA and forces the PKCS1 algorithm (`PSA_ALG_RSA_PKCS1V15_CRYPT` or `PSA_ALG_RSA_PKCS1V15_SIGN(hash_alg)`, and we're either missing a case for `PSA_ALG_RSA_PKCS1V15_SIGN_RAW` or it's happening fortuitously due to the encoding of `psa_algorithm_t`).

Reproducer: https://github.com/gilles-peskine-arm/mbedtls/commit/5e94eb3fc190e6429b1085ede9a031b8f0481f5c in https://github.com/gilles-peskine-arm/mbedtls/tree/pk-rsa-test-padding. When `MBEDTLS_USE_PSA_CRYPTO` is enabled, the new PSS test cases fail.

### PR DESCRIPTION
**Pull request by valeriosetti**
## Description

Resolves #8824

## PR checklist

- [x] **changelog** provided
- [x] **backport** not required - in 2.28, `USE_PSA_CRYPTO` only affects ECDSA, not RSA.
- [x] **tests** provided

### CODE REVIEW

**Review by mpg — COMMENTED** [review]
Looking good, just need to do the same for the other 3 functions (verify, decrypt, encrypt).

**Comment by valeriosetti** [comment]
I'm removing the `need-ci` label because OpenCI fully passed. Internal one is still running after the weekend so I guess it's deadlocked somewhere. I will rebase on top of `development` to solve that. Meanwhile the PR is ready for review.

**Comment by mpg** [comment]
> Internal one is still running after the weekend so I suspect it's deadlocked somewhere

I think it's under maintenance, nothing to worry about.

File: tests/suites/test_suite_pk.function [hunk 1]
```diff
@@ -1055,6 +1082,17 @@ void pk_sign_verify(int type, int curve_or_keybits, int sign_ret, int verify_ret
     TEST_ASSERT(mbedtls_pk_setup(&pk, mbedtls_pk_info_from_type(type)) == 0);
     TEST_ASSERT(pk_genkey(&pk, curve_or_keybits) == 0);
 
+#if defined(MBEDTLS_RSA_C)
+    if (type == MBEDTLS_PK_RSA) {
+        /* Just pick SHA1 here as hashing algorithm as we're more interested
+         * in checking the compatibility between */
```
Comment (mpg) [inline tests/suites/test_suite_pk.function:1088]: Sentence looks incomplete. Anyway, this doesn't look like it's just picking SHA1: `rsa_md_alg` is an input parameter and could hold any value.

**Review by mpg — CHANGES_REQUESTED** [review]
Looks good to me except for an outdated comment (probably about an intermediate version).

File: tests/suites/test_suite_pk.data [hunk 2]
```diff
@@ -384,51 +404,79 @@ pk_ec_test_vec:MBEDTLS_PK_ECKEY:MBEDTLS_ECP_DP_SECP256R1:"0437cc56d976091e5a723e
...
+pk_rsa_decrypt_test_vec:"1253e04dc0a5397bb44a7ab87e9bf2a039a33d1e996fc82a94ccd30074c95df763722017069e5268da5d1c0b4f872cf653c11df82314a67968dfeae28def04bb6d84b1c31d654a1970e5783bd6eb96a024c2ca2f4a90fe9f2ef5c9c140e5bb48da9536ad8700c84fc9130adea74e558d51a74ddf85d8b50de96838d6063e0956":1024:MBEDTLS_RSA_PKCS_V15:MBEDTLS_MD_SHA1:"eecfae81b1b9b3c908810b10a1b5600199eb9f44aef4fda493b81a9e3d84f632124ef0236e5d1e3b7e28fae7aa040a2d5b252176459d1f397541ba2a58fb6599":"c97fb1f027f453f6341233eaaad1d9353f6c42d08866b1d05a0f2035028b9d869840b41666b42e92ea0da3b43204b5cfce3352524d0416a5a441e700af461503":"bbf82f090682ce9c2338ac2b9da871f7368d07eed41043a440d6b6f07454f51fb8dfbaaf035c02ab61ea48ceeb6fcd4876ed520d60e1ec4619719d8a5b8b807fafb8e0a3dfc737723ee6b4b7d93a2584ee6a649d060953748834b2454598394ee0aab12d7b61a51f527a9a41f6c1687fe2537298ca2a8f5946f8e5fd091dbdcb":"11":"d436e99569fd32a7c8a05bbc90d32c49":MBEDTLS_ERR_RSA_INVALID_PADDING
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_pk.data:475]: That's the corrupted ciphertext.
```suggestion
pk_rsa_decrypt_test_vec:"1253e04dc0a5397bb44a7ab87e9bf2a039a33d1e996fc82a94ccd30074c95df763722017069e5268da5d1c0b4f872cf653c11df82314a67968dfeae28def04bb6d84b1c31d654a1970e5783bd6eb96a024c2ca2f4a90fe9f2ef5c9c140e5bb48da9536ad8700c84fc9130adea74e558d51a74ddf85d8b50de96838d6063e0955":1024:MBEDTLS_RSA_PKCS_V15:MBEDTLS_MD_SHA1:"eecfae81b1b9b3c908810b10a1b5600199eb9f44aef4fda493b81a9e3d84f632124ef0236e5d1e3b7e28fae7aa040a2d5b252176459d1f397541ba2a58fb6599":"c97fb1f027f453f6341233eaaad1d9353f6c42d08866b1d05a0f2035028b9d869840b41666b42e92ea0da3b43204b5cfce3352524d0416a5a441e700af461503":"bbf82f090682ce9c2338ac2b9da871f7368d07eed41043a440d6b6f07454f51fb8dfbaaf035c02ab61ea48ceeb6fcd4876ed520d60e1ec4619719d8a5b8b807fafb8e0a3dfc737723ee6b4b7d93a2584ee6a649d060953748834b2454598394ee0aab12d7b61a51f527a9a41f6c1687fe2537298ca2a8f5946f8e5fd091dbdcb":"11":"d436e99569fd32a7c8a05bbc90d32c49":MBEDTLS_ERR_RSA_INVALID_PADDING
```

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
LGTM apart from the documentation and some copypasta in test data.

**Review by gilles-peskine-arm — DISMISSED** [review]
LGTM

**Review by mpg — DISMISSED** [review]
LGTM, thanks!

**Comment by mpg** [comment]
Hmm, actually this is missing a Changelog entry. Should be under "Bugfix". Suggestion:

```
The functions mbedtls_pk_sign(), mbedtls_pk_verify(), mbedtls_pk_decrypt() or mbedtls_pk_encrypt() on a context holding an RSA key that was configured to use PCKS#1 v2.1 (PSS or OAEP), for example by previously calling mbedtls_rsa_set_padding(mbedtls_pk_rsa(...), MBEDTLS_RSA_PKCS1_V21, ...), were incorrectly using PKCS#1 v1.5 when MBEDTLS_USE_PSA_CRYPTO was enabled.
```

**Review by mpg — APPROVED** [review]
LGTM :)

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

File: ChangeLog.d/8824.txt [hunk 3]
```diff
@@ -0,0 +1,6 @@
+Bugfix
+   * The functions mbedtls_pk_sign(), mbedtls_pk_verify(), mbedtls_pk_decrypt()
+     or mbedtls_pk_encrypt() on a context holding an RSA key that was configured
+     to use PCKS#1 v2.1 (PSS or OAEP), for example by previously calling
+     mbedtls_rsa_set_padding(mbedtls_pk_rsa(...), MBEDTLS_RSA_PKCS1_V21, ...),
+     were incorrectly using PKCS#1 v1.5 when MBEDTLS_USE_PSA_CRYPTO was enabled.
```
Comment (gilles-peskine-arm) [inline ChangeLog.d/8824.txt:6]: Non-blocker, but the wording of this entry bothers me a little. It says “this was wrong before”, but it doesn't explicitly say “we've fixed it”. We normally word changelog entries to say “this is the new behavior” as well as “this is why the previous behavior was wrong”.

Also we normally indicate the issue (although I only consider that a blocker when it's user-reported).

So I'd prefer something like

> Fix mbedtls_pk_sign(), mbedtls_pk_verify(), mbedtls_pk_decrypt() or mbedtls_pk_encrypt() on non-opaque RSA keys to honor the padding mode in the RSA context. Before they always used PKCS#1 v1.5 even when the RSA context was configured for PKCS#1 v2.1 (PSS/OAEP). Fixes #8824.

File: ChangeLog.d/8824.txt [hunk 4]
```diff
@@ -0,0 +1,6 @@
+Bugfix
+   * Fix mbedtls_pk_sign(), mbedtls_pk_verify(), mbedtls_pk_decrypt() or
```
Comment (gilles-peskine-arm) [inline ChangeLog.d/8824.txt:2]: Sorry, one niggle I didn't spot earlier:
```suggestion
   * Fix mbedtls_pk_sign(), mbedtls_pk_verify(), mbedtls_pk_decrypt() and
```

File: ChangeLog.d/8824.txt [hunk 5]
```diff
@@ -0,0 +1,6 @@
...
+     mbedtls_pk_encrypt() on non-opaque RSA keys to honor the padding mode in
+     the RSA context. Before they always used PKCS#1 v1.5 even when the RSA
+     context was configured for PKCS#1 v2.1 (PSS/OAEP). Fixes #8824.
```
Comment (mpg) [inline ChangeLog.d/8824.txt:5]: Thanks for improving the format, but I think the description of the old behaviour is incomplete: the issue was only when MBEDTLS_USE_PSA_CRYPTO was enabled. Suggestion: "Before, if MBEDTLS_USE_PSA_CRYPTO was enabled, they always used..."

Comment (gilles-peskine-arm) [reply ChangeLog.d/8824.txt:5]: Oops...

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

**Review by mpg — APPROVED** [review]
LGTM, thanks!

**Pull request merged**